File: go/test/provider_settings_write_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/cloudsmith-sdk/go"
	"github.com/voxgig-sdk/cloudsmith-sdk/go/core"

	vs "github.com/voxgig-sdk/cloudsmith-sdk/go/utility/struct"
)

func TestProviderSettingsWriteEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ProviderSettingsWrite(nil)
		if ent == nil {
			t.Fatal("expected non-nil ProviderSettingsWriteEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := provider_settings_writeBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "provider_settings_write." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set CLOUDSMITH_TEST_PROVIDER_SETTINGS_WRITE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		providerSettingsWriteRef01Ent := client.ProviderSettingsWrite(nil)
		providerSettingsWriteRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "provider_settings_write"}, setup.data), "provider_settings_write_ref01"))
		providerSettingsWriteRef01Data["org_id"] = setup.idmap["org01"]

		providerSettingsWriteRef01DataResult, err := providerSettingsWriteRef01Ent.Create(providerSettingsWriteRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		providerSettingsWriteRef01Data = core.ToMapAny(providerSettingsWriteRef01DataResult)
		if providerSettingsWriteRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// UPDATE
		providerSettingsWriteRef01DataUp0Up := map[string]any{
			"org_id": setup.idmap["org_id"],
		}

		providerSettingsWriteRef01MarkdefUp0Name := "mapping_claim"
		providerSettingsWriteRef01MarkdefUp0Value := fmt.Sprintf("Mark01-provider_settings_write_ref01_%d", setup.now)
		providerSettingsWriteRef01DataUp0Up[providerSettingsWriteRef01MarkdefUp0Name] = providerSettingsWriteRef01MarkdefUp0Value

		providerSettingsWriteRef01ResdataUp0Result, err := providerSettingsWriteRef01Ent.Update(providerSettingsWriteRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		providerSettingsWriteRef01ResdataUp0 := core.ToMapAny(providerSettingsWriteRef01ResdataUp0Result)
		if providerSettingsWriteRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if providerSettingsWriteRef01ResdataUp0[providerSettingsWriteRef01MarkdefUp0Name] != providerSettingsWriteRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", providerSettingsWriteRef01MarkdefUp0Name, providerSettingsWriteRef01ResdataUp0[providerSettingsWriteRef01MarkdefUp0Name])
		}

	})
}

func provider_settings_writeBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "provider_settings_write", "ProviderSettingsWriteTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read provider_settings_write test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse provider_settings_write test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"provider_settings_write01", "provider_settings_write02", "provider_settings_write03", "org01", "org02", "org03", "openid_connect01", "openid_connect02", "openid_connect03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("CLOUDSMITH_TEST_PROVIDER_SETTINGS_WRITE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"CLOUDSMITH_TEST_PROVIDER_SETTINGS_WRITE_ENTID": idmap,
		"CLOUDSMITH_TEST_LIVE":      "FALSE",
		"CLOUDSMITH_TEST_EXPLAIN":   "FALSE",
		"CLOUDSMITH_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["CLOUDSMITH_TEST_PROVIDER_SETTINGS_WRITE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add org_id alias for update test.
	if idmapResolved["org_id"] == nil {
		idmapResolved["org_id"] = idmapResolved["org01"]
	}

	if env["CLOUDSMITH_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["CLOUDSMITH_APIKEY"],
			},
			extra,
		})
		client = sdk.NewCloudsmithSDK(core.ToMapAny(mergedOpts))
	}

	live := env["CLOUDSMITH_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["CLOUDSMITH_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: js/test/entity/provider_settings_write/ProviderSettingsWriteEntity.test.js

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

const Path = require('node:path')
const Fs = require('node:fs')

const { test, describe } = require('node:test')
const assert = require('node:assert')


const { CloudsmithSDK, BaseFeature, stdutil, config } = require('../../..')

const {
  envOverride,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
} = require('../../utility')


describe('ProviderSettingsWriteEntity', async () => {

  test('instance', async () => {
    const testsdk = CloudsmithSDK.test()
    const ent = testsdk.ProviderSettingsWrite()
    assert(null != ent)
  })


  test('basic', async () => {

    const setup = basicSetup()
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select


    // CREATE
    const provider_settings_write_ref01_ent = client.ProviderSettingsWrite()
    let provider_settings_write_ref01_data = setup.data.new.provider_settings_write['provider_settings_write_ref01']
    provider_settings_write_ref01_data['org_id'] = setup.idmap['org01']

    provider_settings_write_ref01_data = await provider_settings_write_ref01_ent.create(provider_settings_write_ref01_data)
    assert(null != provider_settings_write_ref01_data)


    // UPDATE
    const provider_settings_write_ref01_data_up0 = {}
    provider_settings_write_ref01_data_up0 ['org_id'] = setup.idmap['org_id']

    const provider_settings_write_ref01_markdef_up0 = { name: 'mapping_claim', value: 'Mark01-provider_settings_write_ref01_' + setup.now }
    provider_settings_write_ref01_data_up0 [provider_settings_write_ref01_markdef_up0.name] = provider_settings_write_ref01_markdef_up0.value

    const provider_settings_write_ref01_resdata_up0 = await provider_settings_write_ref01_ent.update(provider_settings_write_ref01_data_up0)
    assert(null != provider_settings_write_ref01_resdata_up0)

    assert(provider_settings_write_ref01_resdata_up0[provider_settings_write_ref01_markdef_up0.name] === provider_settings_write_ref01_markdef_up0.value)


  })
})



function basicSetup(extra) {
  // TODO: fix test def options
  const options = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname,
      '../../../../.sdk/test/entity/provider_settings_write/ProviderSettingsWriteTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = CloudsmithSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['provider_settings_write01','provider_settings_write02','provider_settings_write03','org01','org02','org03','org01','org02','org03','openid_connect01','openid_connect02','openid_connect03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'CLOUDSMITH_TEST_PROVIDER_SETTINGS_WRITE_ENTID': idmap,
    'CLOUDSMITH_TEST_LIVE': 'FALSE',
    'CLOUDSMITH_TEST_EXPLAIN': 'FALSE',
    'CLOUDSMITH_APIKEY': 'NONE',
  })

  idmap = env['CLOUDSMITH_TEST_PROVIDER_SETTINGS_WRITE_ENTID']

  if ('TRUE' === env.CLOUDSMITH_TEST_LIVE) {
    client = new CloudsmithSDK(merge([
      {
        apikey: env.CLOUDSMITH_APIKEY,
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.CLOUDSMITH_TEST_EXPLAIN,
    now: Date.now(),
  }

  return setup
}
  
